exploration/number-selection.md

```diff
@@ -116,20 +181,52 @@ but can cause problems in target locales that the original developer is not cons
 > [*] {{You have {$var} chances remaining}}
 > ```
 
-Additional options such as `minimumFractionDigits` and others already supported by `:number`
-should also be supported.
+### Other Numeric Selector Functions
+
+The function `:plural` operates identically to `:number`, except:
+- the `option` of `select` is not valid
+- the `select` type of this selector is always `plural`
+- the function is not valid as a formatter
+
+The function `:ordinal` operates identically to `:number`, except:
+- the `option` of `select` is not valid
+- the `select` type of this selector is always `ordinal`
+  (This function is valid as a formatter)
+
+The function `:integer` operates identically to `:number`, except:
+- the `minimumFractionDigits` and `maximumFranctionDigits` options
+  default to `0` even for floating point numeric types
+> [!NOTE]
+> The fraction digits can be overridden for the purposes of formatting
+> and selection, but the result must be that any fractional part of
+> the operand is reduced to `0`
 
-If PR [#532](https://github.com/unicode-org/message-format-wg/pull/532) is accepted,
-also add the following `<alias>` definitions to `<function name="number">`:
+### Plural/Ordinal Keywords
+The _plural/ordinal keywords_ are: `zero`, `one`, `two`, `few`, `many`, and
+`other`.
+The value `other` is both a valid output of plural or ordinal rule functions
+and a default value.
 
-```xml
-<alias name="plural" supports="match">
-	<setOption name="select" value="plural"/>
-</alias>
-<alias name="ordinal" supports="match">
-	<setOption name="select" value="ordinal"/>
-</alias>
+### Determining Exact Literal Match
+A numeric literal key value exactly matches the resolved value of an operand as follows:
+
+> this is still in progress, but basically has to consider:
+
+- signs match
+- minimum fraction digits match
+- scientific notation?
+- other formatting gorp doesn't apply (e.g. currency or measure unit)
+- no grouping
```

Comment:
Given that we have `:string` separately, `:number` should so numerical comparison:
```suggestion
To determine if a numeric literal key value
exactly matches the resolved value of an operand,
the parsed numerical value represented by the key
is compared to the numerical value of the _operand_.
Selector options such as `minimumFractionDigits` have no effect on this comparison.
```

With that, the keys `1.0` and `1` would both match a number 1, with equal weight.

Reply:
Floating point vs. integer representation might introduce some corner cases. I'm not sure about "equal weight" either. Our specification requires a list at the moment, so we should provide a tie-break so that the order is deterministic.

We also have to be cautious. Suppose the selector is `{$num :number maximumFractionDigits=1}`, the value in the operand is `1.0001`, and the key is `1.0`. This should probably be a match, right? Even though the numerical values are not (quite) equal?

Reply:
A couple of points:
- As @aphillips notes, many languages have different behavior depending on the presence or absence of trailing zero fraction digits, so those must be maintained. That is, |1|, |1.0|, |1.00|, etc.
- The interaction between max/min digits and significant digits is surprisingly complicated. Describing the precise behavior should not be attempted in this spec.

Reply:
No, that should not match as the numerical values are not equal.

While I agree that the behaviour around non-integer values may be a little surprising, as far as I know they are only theoretical, and should not influence the overall design unduly. Or are there any real-world examples of current messages that need to work around the limitations of MF1, Fluent, or any other syntax that supports selection on exact numerical values? I have searched a bit, but have not found any.

Reply:
@eemeli I didn't take you text verbatim, but I took the idea into account. Check it out.

@macchiati Thanks. The matching behavior for literal keys I don't think is affected by this, but certainly the plural/ordinal rule computation is. I think we've accounted for that. 

I agree that we don't want to get into signfiicant digits and such, although I am worried about developers not being able to figure out why setting the number of digits or modifying the rounding mode doesn't cause operand values to match certain key literals (as @eemeli suggests). That is, why doesn't `{|1.00000001| :number maximumFractionDigits=1}` match `1.0` (since it formats as "1.0" or the equivalent)?

Reply:
@macchiati I think we understand.

Part of the problem here is that your example doesn't annotate `$averageWater`. It would either print `1` as an integer or `1.0`, but not both. If `$averageWater` is annotated to produce a fraction part, plural rules would handle getting the right keyword (including for `1.0`). You wouldn't write the literal `|1|`.

Let me rewrite your example a bit:

```
.match {$averageWater :number minimumFractionDigits=1}
one    {{This never fires in English; {$averageWater} liter per person}}
*      {{This can fire in English;    {$averageWater} liters per person}}
1      {{This matches;                {$averageWater} liter per person}}
1.0    {{This also matches;           {$averageWater} liters per person}}
```

The problem with this message and maybe @eemeli's suggestion is that `1` matches because it has the value `1.0`. This makes writing non-grammatical messages perhaps slightly more likely. It's also unclear what user expectations would be for such a message.

My suggestion would be that the key has to match the significant digits in the selector function in order to match. This might involve trimming or rounding the operand. Describing this is much more complicated that just parsing the key and doing `.equals(operand)` on it. It would probably be more like "serialize the operand according to the JSON rules to the number of significant fraction digits specified in the numeric selector function; The key and operand are an exact match if the resulting string is code point equal to the key".

Is this significantly better than the parse solution? It might be if floating point or fractional values are commonly used as literal keys. My experience, though, is that most literal keys are integers in integer applications. Should we provide an affordance for these more common cases? Or is full control worth the extra cost and complexity?

Reply:
I think the the most robust approach would be the following description (I'm not completely clear if this is just restating what your words were or not).

A literal matches a variable using {$variable :number ...}, iff after formatting the literal using the same settings as the $variable has, there is an exact string match. 

 It can clearly be optimized in production, but if we state it like this it establishes the principle. It also means that we can detect that 1 are equivalent: will either both match or neither match, which can be detected statically.

Reply:
Let me give it a shot at wording. I think this is the better approach.

Reply:
MF2 has _multiple selection_ and that might work better than trying to shove everything into a single selector:

```
.local $offset = {$numGuides :number offset=1}
.match {$numGuests :number} {$offset}
0   *   {{No guests were invited}}
1   *   {{Only {$guest1} was invited}}
*   one {{Only {$guest1} guest and {$offset} more guest}}
*   *   {{Only {$guest1} and {$offset} more were invited}}
```

(Note that other languages would write `one one`/`one few`/`one *`/etc. messages to fit grammatical needs. The keys `0 *` and `1 *` act as guardrails against the weird behavior of `$offset`)

Reply:
Good point; that would involve just one additional operand ('offset', or maybe 'minus'). Not terse, but works.

```
.local $offset = {$numGuests :number offset=1}
.match {$numGuests :number} {$offset}
0   *   {{No guests were invited}}
1   *   {{Only {$guest1} was invited}}
*   one {{{$guest1} and {$offset} more guest was invited}}
*   *   {{{$guest1} and {$offset} more guests were invited}}
```